### PR TITLE
[Sema] Avoid opening base requirements in static/init calls

### PR DESCRIPTION
**Pull request by xedin**
Avoid re-introducing the same base generic requirements
into constraint system multiple times for static member
and init calls, the problem is related to the fact that
constraint generation is opening base type and solver is
opening function overload choice and signatures have
intersecting requirements, e.g.

```swift
class A<T: Hashable> {
  init(_: [T]) {}
}

_ = A([0, 42])
```

Since call has reference to a type `A<T>.Type` by the time we
attempt to open reference to initializer `(A<T>) -> ([T]) -> A<T>`
`T` would already be opened and all of its generic requirements
would be registered in the constraint system.

Alternative approach to https://github.com/apple/swift/pull/18326

### CODE REVIEW

**Comment by xedin** [comment]
I'm not sure what would be the best way to test changes like these so open to suggestions...

**Comment by xedin** [comment]
I'll try to do `-debug-constraints` and get it through `FileCheck`.

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please smoke test compiler performance

**Comment by xedin** [comment]
@swift-ci please smoke test compiler performance

**Comment by xedin** [comment]
Source Compatibility failure is a `UPASS` of `CoreStore`.

File: lib/Sema/ConstraintSystem.cpp [hunk 1]
```diff
@@ -441,11 +442,9 @@ Type ConstraintSystem::openUnboundGenericType(UnboundGenericType *unbound,
 
   // Open up the generic type.
   openGeneric(unboundDecl->getInnermostDeclContext(),
-              unboundDecl->getDeclContext(),
-              unboundDecl->getGenericSignature(),
-              /*skipProtocolSelfConstraint=*/false,
-              locator,
-              replacements);
+              unboundDecl->getDeclContext(), unboundDecl->getGenericSignature(),
+              /*skipProtocolSelfConstraint=*/false, locator, replacements,
```
Comment (slavapestov) [inline lib/Sema/ConstraintSystem.cpp]: It might be worth seeing if skipProtocolSelfConstraint can be removed.

**Pull request closed without merging**